src/webgpu/shader/execution/expression/call/builtin/texture_utils.ts

```diff
@@ -222,6 +262,73 @@ export function expected<T extends Dimensionality>(
   }
 }
 
+/**
+ * The software version of a texture builtin (eg: textureSample)
+ * Note that this is not a complete implementation. Rather it's only
+ * what's needed to generate the correct expected value for the tests.
+ */
+export function softwareTextureRead<T extends Dimensionality>(
+  call: TextureCall<T>,
+  texture: Texture,
+  sampler: GPUSamplerDescriptor,
+  size: [number, number]
+): PerTexelComponent<number> {
+  assert(call.ddx !== undefined);
+  assert(call.ddy !== undefined);
+  const rep = kTexelRepresentationInfo[texture.texels[0].format];
+  const texSize = reifyExtent3D(texture.descriptor.size);
+  const textureSize = [texSize.width, texSize.height];
+
+  const ddx: readonly number[] = typeof call.ddx === 'number' ? [call.ddx] : call.ddx;
+  const ddy: readonly number[] = typeof call.ddy === 'number' ? [call.ddy] : call.ddy;
+  const sDdx = ddx.map((v, i) => (v * textureSize[i]) / size[i]);
+  const sDdy = ddy.map((v, i) => (v * textureSize[i]) / size[i]);
+  const dotDDX = dotProduct(sDdx, sDdx);
+  const dotDDY = dotProduct(sDdy, sDdy);
+  const deltaMax = Math.max(dotDDX, dotDDY);
+
+  // MAINTENANCE_TODO: handle texture view baseMipLevel and mipLevelCount?
+  const mipLevel = 0.5 * Math.log2(deltaMax);
+  const mipLevelCount = texture.texels.length;
+  const maxLevel = mipLevelCount - 1;
+
+  switch (sampler.mipmapFilter) {
+    case 'linear': {
+      const clampedMipLevel = clamp(mipLevel, { min: 0, max: maxLevel });
+      const baseMipLevel = Math.floor(clampedMipLevel);
+      const nextMipLevel = Math.ceil(clampedMipLevel);
+      const t0 = expected<T>(call, texture, sampler, baseMipLevel);
```

Comment:
nit: perhaps rename `expected` to something more descriptive

Reply:
done